Publish APIs

jmendeza opened the pull request:
https://github.com/craftercms/craftercms/issues/6766
Publish APIs

https://github.com/craftercms/craftercms/issues/6398

jmendeza committed:
New publish APIs

jmendeza committed:
API rename to publish/dependencies. Tweaks to publish package request

jmendeza committed:
requestApproval parameter for publish API

jmendeza committed:
Add publishAll flag to publish API. Add submitter param filter to get packages API

jmendeza committed:
repository/log API

jmendeza committed:
API specs: publish trigger, package retry, package clone

jmendeza committed:
Publishing progress related responses

jmendeza committed:
Remove non-implemented/removed APIs from spec

jmendeza committed:
Deprecate old publish APIs

jmendeza committed:
New Publisher db tables. Remove old tables.

jmendeza committed:
Remove old publisher-related code

jmendeza committed:
Remove item previous_path

jmendeza committed:
Make published repo bare
Initial publish updates

jmendeza committed:
Studio API tweaks

jmendeza committed:
New publish API methods skeleton. Publish services refactor

jmendeza committed:
Publish all package submission

jmendeza committed:
PublishItem DAO
Regular publish package builder

jmendeza committed:
PublishDAO updates
Regular publishing package submission

jmendeza committed:
ItemTargetDAO. Publishing package submission updates

jmendeza committed:
Include children in publishing package when requested

jmendeza committed:
Publish-all for any request if site has not been published. API tweaks

jmendeza committed:
API 1 commits and bulk-golive tweaked and marked as deprecated. Bulk publish root becomes publish-all

jmendeza committed:
Publish calculation package tweaks: get soft deps of children paths

jmendeza committed:
Publish APIs tweaks. Publish dependencies calculation

sumerjabri commented on src/main/api/studio-api.yaml:
```diff
@@ -6579,12 +6568,125 @@ paths:
+                notifySubmitter:
+                  type: boolean
+                  description: notify the submitter when request is approved/rejected
+                publishAll:
```
Does this trump the `PublishPackageRequest`? Feels a bit off. Let's discuss this a bit.

jmendeza replied:
`PublishPackageRequest` params are not required. Notice that PublishPackageRequest params are flattened in this request body, it can be as simple as:
```
{
    "siteId": "ed1",
    "publishingTarget": "live",
    "publishAll": true
}
```
or contain the list of paths
```
{
    "siteId": "ed1",
    "publishingTarget": "live",
    "paths": [
        {
            "path": "/site/website/articles/",
            "includeChildren": true,
            "includeSoftDeps": true
        }
    ]
}
```

sumerjabri replied:
I assume combining the two results will make PublishAll win?

jmendeza replied:
Yes, publish all flag will take precedence

sumerjabri commented on src/main/api/studio-api.yaml:
```diff
@@ -6579,12 +6568,125 @@ paths:
+                requestApproval:
+                  type: boolean
+                  description: submit the request for approval
+                notifySubmitter:
```
Why do we have this? Would there be a case where we don't?

jmendeza replied:
Removed (the `notifySubmitter` param) as discussed

sumerjabri commented on src/main/api/studio-api.yaml:
```diff
@@ -6829,7 +6938,199 @@ paths:
+        '500':
+          $ref: '#/components/responses/InternalServerError'
+
+  /api/2/publish/package/{packageId}/retry:
```
Let's remove this API per our conversation.

jmendeza replied:
Done

sumerjabri commented on src/main/api/studio-api.yaml:
```diff
@@ -10327,6 +10298,92 @@ components:
+              processed: 100
+              total: 100
+              errors: false
+            - name: "Live publish"
```
Let's add staging publish here (we can indicate that it's optional)

sumerjabri commented on src/main/java/org/craftercms/studio/api/v1/service/deployment/DeploymentService.java:
```diff
@@ -73,7 +39,7 @@ List<ContentItemTO> getScheduledItems(String site, String sort, boolean ascendin
      * @throws ServiceLayerException exception is case of en error
      */
-    void bulkGoLive(String site, String environment, String path, String comment) throws ServiceLayerException;
+    long bulkGoLive(String site, String environment, String path, String comment) throws ServiceLayerException, AuthenticationException;
```
Rename to bulkPublish?

jmendeza replied:
@sumerjabri I would prefer to leave it like that so it matches the deprecated v1 API (`api/1/deployment/bulk-golive.json`) for the rest of its lifetime. This piece of code will be removed anyway when `bulk-golive` API is retired.

jmendeza committed:
Merge branch 'develop' of https://github.com/craftercms/studio into feature/6766

jmendeza committed:
Remove 'notifySubmitter' option from APIs and publish package

jmendeza committed:
Drop the 'publish package retry' API in favor or using publish/package/{packageId}/clone'

jmendeza committed:
Dependency services refactor

jmendeza committed:
Hard dependencies. API spec tweaks

sumerjabri commented on src/main/java/org/craftercms/studio/api/v2/dal/publish/PublishItem.java:
```diff
@@ -0,0 +1,122 @@
+    protected Action action;
+    protected boolean userRequested;
+    protected State state;
+    protected String error;
```
Is this where we would add the new fields?

jmendeza replied:
The item metadata/cached fields like label, systemType, etc. Yes.

jmendeza committed:
Workflow events for publish submission

jmendeza committed:
Transactional publishing package+items insert

jmendeza committed:
Simplify ItemDAO result mappers. Update queries to remove references to old workflow table

jmendeza committed:
Event-based publishing
Initial publish

sumerjabri commented on src/main/java/org/craftercms/studio/api/v1/service/deployment/DeploymentService.java:
```diff
@@ -73,36 +37,26 @@ List<ContentItemTO> getScheduledItems(String site, String sort, boolean ascendin
      * Publish given commit IDs on given environment for given site
      * @param site site id to use for publishing
      * @param environment environment to use for publishing
      * @param commitIds commit IDs to publish
```
What's the returned `long` for? Can we update the JavaDoc?

jmendeza replied:
Added to the javadoc. The returned long is the id of the create publish package

sumerjabri commented on src/main/java/org/craftercms/studio/api/v1/service/deployment/DeploymentService.java:
```diff
@@ -73,36 +37,26 @@ List<ContentItemTO> getScheduledItems(String site, String sort, boolean ascendin
      * @throws ServiceLayerException exception is case of en error
      */
-    void bulkGoLive(String site, String environment, String path, String comment) throws ServiceLayerException;
+    long bulkGoLive(String site, String environment, String path, String comment) throws ServiceLayerException, AuthenticationException;
```
Let's indicate what the `long` is for in the JavaDoc.

jmendeza replied:
Added to the javadoc. The returned long is the id of the create publish package

jmendeza committed:
Initial publish tweaks. Publish all (WIP)

jmendeza committed:
Do NOT publish-all when publishing the root

jmendeza committed:
javadoc tweaks

jmendeza committed:
Hard dependency calculation tweaks. Delete publish_package records on new site delete

jmendeza committed:
Initial publish tweaks. last_published_on update

jmendeza committed:
ItemTargetDAO operations

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Add packageType to PublishPackage. Tweaks to package submission. Simplify and consolidate package creation

jmendeza committed:
Populate item_publish_item table
Set system processing while publishing

jmendeza committed:
Publish all processing (WIP)

jmendeza committed:
Move PublishItemTO implementations to impl package

jmendeza committed:
Publish all processing and commit-tree

jmendeza committed:
'item-list' publish package processing (WIP). DB updates

jmendeza committed:
updateRef method

jmendeza committed:
Publish item-list: update index and write tree

jmendeza committed:
Publish item-list: update-index for deletes

jmendeza committed:
Tweaks to publish package submission

jmendeza committed:
Tweaks to publish package processing to reuse code

jmendeza committed:
Refactor ContentRepository and related interfaces

jmendeza committed:
Fix repo path when publishing blobs. Fix tests.

jmendeza committed:
Solve warnings

jmendeza committed:
BlobAwareContentRepository publish unit tests

jmendeza committed:
Transaction support. Utility function to run a piece of code in a transaction

jmendeza committed:
Publisher: updated DB to support including more information about errors and package progress

jmendeza committed:
Handle enabling staging after initial publish

jmendeza committed:
Publisher: publish each target in its own transaction

jmendeza committed:
PublishPackageTO to simplify target-specific operations on the PublishTarget

jmendeza committed:
Fix publish commit comment

jmendeza committed:
Fix blob publishing when there are no deletes (or no updates)

jmendeza committed:
Publisher states DAO updates to be able to set on/off specific bits

jmendeza committed:
Fix PublishPackage mybatis mapping of siteId property

jmendeza committed:
Publish progress states updates. Fix test

jmendeza committed:
PublishPackage completed with errors state

jmendeza committed:
Update '/workflow/request_publish' API to work with the new system. API spec tweaks

jmendeza committed:
Update items states when submitting publish package

jmendeza committed:
Cancel outstanding packages after a publish all: fix site_id param and update to states bitmap

jmendeza committed:
Tweaks to item and package state updates

jmendeza committed:
Publish package publishedOn moved to package level processing

jmendeza committed:
Publish error handling

jmendeza committed:
Publish transaction error logging tweaks and cleanup

jmendeza committed:
Fix new items not retrieving the ids when inserting list

jmendeza committed:
Fix hard-dependencies site filtering

jmendeza committed:
Result reporting on AWS copy list of blobs

jmendeza committed:
Missing error log

jmendeza committed:
Initialize stage item_target records on first staging publish. DB rename old_path -> previous_path

jmendeza committed:
Update item_target previous_path on item move

jmendeza committed:
Clear published previous_path after a package is complete

jmendeza committed:
Remove unused code

jmendeza committed:
Break down large query into smaller pieces

jmendeza committed:
Migrate to new deleteSite V2 service

jmendeza committed:
Duplicate site to duplicate item_target table

jmendeza committed:
Fix Enabled/disable publisher API for new publisher

jmendeza committed:
Reorganize WorkflowService impl/internal impl

jmendeza committed:
Remove unused code

jmendeza committed:
ContentItemTO metadata for compatibility. Fix edit form with new publisher updates. Tweaks to Item DAL to parameterize packageState and approvalState values

jmendeza committed:
Get childen by path query to parameterize packageState and approvalState values, fixed packageState filtering

jmendeza committed:
Fix available actions when 'require peer review' is enabled

jmendeza committed:
Added logging for published items

jmendeza committed:
Fix initial publish package state

jmendeza committed:
Publisher audit/activity tweaks

jmendeza committed:
Publisher-related components synchronization/locking

jmendeza committed:
Content services minor refactor

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Fix SitesService tests

jmendeza committed:
Publish DELETE operations

jmendeza committed:
Prevent processedCommits insert attempt for global site

jmendeza committed:
Fix clearPreviousPathForCompletePackage to filter by site

jmendeza committed:
Removed unused/obsolete code

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Remove unnecessary sync-from-repo task locking

jmendeza committed:
Remove unused dependencies code

jmendeza committed:
Renamed flag to keep track of renamed content even when previous_path should be cleared

jmendeza committed:
Remove .keep files from published

jmendeza committed:
Code cleanup. Workflow cancellation method wiring

jmendeza committed:
Lock packages when publishing. Cancel workflow on sync-from-repo

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Fail content write when item is in publish queue

jmendeza committed:
Check if there are content children in the publishing queue when moving/deleting

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Check if there are content children in the publishing queue when renaming

jmendeza committed:
Preserve IN_WORKFLOW/SCHEDULED state bits after on package complete if there are other packages affecting the item

jmendeza committed:
API spec updates to include 409 response for conflict with publish queue

jmendeza committed:
Cancel publishing packages API update to work with new packages with numeric IDs

jmendeza committed:
Update cancel workflow by path to work with new cancel package and item states update methods

jmendeza committed:
Code cleanup

jmendeza committed:
Preserve DESTINATION state bit on package complete/cancel if there are other packages affecting the item

jmendeza committed:
Move cancel packages by path to publish service

jmendeza committed:
Retire old unused delete method

jmendeza committed:
Handle publish package not found scenario

jmendeza committed:
Fail pull_from_remote when would-be-merged changes conflict with the publish queue

jmendeza committed:
Enable/disable publisher API v2

jmendeza committed:
Reuse updatePackageState method to cancel packages in publishDao

jmendeza committed:
Audit package cancel on sync from repo. Audit log tweaks

jmendeza committed:
Simplify PublishPackage updates in PublishDao

jmendeza committed:
Sync from repo cancel packages updates

jmendeza committed:
Make cancel publishing package API accept a single package id. Add comment to cancel package API.  Handle invalid package state scenario

jmendeza committed:
Reject publish package API. Move reject/cancel to workflow service. Approve publish pacakge API

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Trigger notifications on workflow events

jmendeza committed:
Package submission workflow notification

jmendeza committed:
Package rejection/approval workflow notification

jmendeza committed:
Workflow notification deprecated variables

jmendeza committed:
Workflow approve to support editing package scheduled date

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Fix NPE on notification message

jmendeza committed:
Add UPDATE action to publish item

jmendeza committed:
ItemDAO.getItemStates bulk operation to get item states by path

jmendeza committed:
Dashboard publish history and publish stats API

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Remove deprecated /api/2/publish/history API

jmendeza committed:
Fix package details total item count

jmendeza committed:
Remove old workflow/affected_paths API, added new workflow/affected_packages

jmendeza committed:
List/get/details APIs for publishing package

jmendeza committed:
Publish/Workflow API updates to standardize API Urls

jmendeza committed:
Rename publish DAL methods to be consistent. Cleanup and remove redundant operations

jmendeza committed:
Change package id parameter type

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Publisher-related API spec updates

jmendeza committed:
Publish progress report API updates

jmendeza committed:
Progress report tweaks

jmendeza committed:
Deprecate dashboard publish package-listing APIs in favor of new /api/2/publish/{siteId}/package/{packageId}

jmendeza committed:
Extract query to filter packages by paths

jmendeza committed:
Fix publishing multiple packages for a site

jmendeza committed:
New publish package filtering API

jmendeza committed:
Code cleanup. Remove unused methods and old publish APIs

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Remove old column previous_path index declaration

jmendeza commented:
Notice this PR depends on https://github.com/craftercms/commons/pull/398

jmendeza committed:
Add 'user' back to WorkflowEvent

sumerjabri committed:
Update GitContentRepository.java

jmendeza committed:
Notification config updates for built-in blueprints

jmendeza committed:
Commented code cleanup

jmendeza committed:
Merge branch 'feature/6766' of github.com:jmendeza/studio into feature/6766

jmendeza committed:
Remove commented code

jmendeza committed:
Fix getPackageForItems query to return packages for any target

jmendeza committed:
Merge remote-tracking branch 'upstream/develop' into feature/6766

jmendeza committed:
Merge branch 'develop' into feature/6766

jmendeza committed:
Tweaks from feedback:
  - Move s3 related errors to AWS section in ApiResponse
  - Language tweaks in logs/error messages
  - "publishing package" -> "publish package"

jmendeza committed:
Merge branch 'feature/6766' of github.com:jmendeza/studio into feature/6766

jmendeza committed:
ApiResponse codes for errors related to AWS S3
